### PR TITLE
Adding support for GPT OSS MoE to YALIS

### LINKED ISSUE #14: Support more models
Gemma models require non-standard variants of attention like softmax logit capping. This is currently not supported in yalis.

### PR DESCRIPTION
**Pull request by hoffmann-muki**
- Replicated the GPT OSS MoE checkpoint, ensuring parity with the OpenAI implementation

### CODE REVIEW

File: yalis/external/checkpoints/openai/gpt-oss-20b
Comment (prajwal1210) [inline yalis/external/checkpoints/openai/gpt-oss-20b:1]: Checkpoints should not be commited. These are big and will pollute the repo. Remove from git history.

File: tests/basic_correctness/conftest.py [hunk 1]
```diff
@@ -140,7 +140,7 @@ def yalis_engine(model_id, dtype, attn_backend):
         max_length_of_generated_sequences=2048,
         top_p=0.0,
         temperature=0.0,
-        tp_dims=None,
+        tp_dims=(2,1,1),
```
Comment (prajwal1210) [inline tests/basic_correctness/conftest.py]: Using None is better, it automatically uses the (N_GPUS, 1, 1) when run on N_GPUs. Revert this change.

File: yalis/external/checkpoints/openai/gpt-oss-20b
Comment (hoffmann-muki) [reply yalis/external/checkpoints/openai/gpt-oss-20b:1]: I've removed it

File: tests/basic_correctness/conftest.py (hunk 1 above)
Comment (hoffmann-muki) [reply tests/basic_correctness/conftest.py]: reverted back to None

File: examples/infer.py [hunk 2]
```diff
@@ -74,14 +65,14 @@
...
+    model_config = ModelConfig(model_name="gpt-oss-20b", model_path=model_id, precision="bf16")
     inference_config = InferenceConfig(
         max_batch_size=MAX_BATCH_SIZE,
         max_length_of_generated_sequences=1024,
         top_p=0.80,
-        temperature=1.0,
+        temperature=0.0,
         tp_dims=None,
-        attention_backend="flash",
+        attention_backend="sdpa",
```
Comment (prajwal1210) [inline examples/infer.py]: Have we not got it working with flash? We really need flash-attn support

File: tests/basic_correctness/conftest.py
Comment (prajwal1210) [inline tests/basic_correctness/conftest.py:1]: We should not change the default test. We should add another .ini file if needed and run that. This should not be pushed.

File: tests/basic_correctness/flash.ini
Comment (prajwal1210) [inline tests/basic_correctness/flash.ini:1]: Revert

File: tests/basic_correctness/sdpa.ini
Comment (prajwal1210) [inline tests/basic_correctness/sdpa.ini:1]: Revert

File: yalis/tensor_parallel/linear.py [hunk 3]
```diff
@@ -411,12 +411,28 @@ def _is_full_weight_matrix(self, weight):
         )
 
     def _is_sharded_weight_matrix(self, weight):
-        return (
-            weight.ndim == 2
-            and weight.size(0) == self.local_out_features
+        # Accept multiple on-disk sharding layouts:
```
Comment (prajwal1210) [inline yalis/tensor_parallel/linear.py]: Why is this needed? I did not understand why the existing thing won't work?

**Review by prajwal1210 — COMMENTED** [review]
A bunch of cleanup needed. The existing tests should not change and should not have local paths. We should be creating a new .ini file if we feel we need to add tests specific to this model. Also a few questions on why certain things are changing.

File: examples/infer.py (hunk 2 above)
Comment (hoffmann-muki) [reply examples/infer.py]: The openAI implementation uses sdpa. Also there was no support for other attention backends. I also did some research on why flash attention was not supported and realized it is because gpt oss attention uses sinks when computing attention and drops them later on. Flash attention does not support logit modifications between Q@K and softmax, and since flash attention computes Q@K, mask, softmax, and weighted sum all inside a single CUDA kernel, adding a sink column after Q@K would mean we have to rewrite the kernel, and we risk breaking the kernel fusion and removing the entire performance benefit

Comment (hoffmann-muki) [reply examples/infer.py]: Also to support flex attention we a dynamic per-head sink column that approximates the sink behavior in the openAI implementation, so there will be some loss of accuracy. That's why i just stuck to using sdpa

Comment (prajwal1210) [reply examples/infer.py]: I see. We should add todo here to find out faster kernels. I see flash infer has a separate kernel for the model. We might need to switch out to using something else.

File: examples/infer.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 
 if __name__ == "__main__":
     # Model ID from Hugging Face
-    model_id = "meta-llama/Llama-3.1-8B-Instruct"
+    model_id = "yalis/external/checkpoints/openai/gpt-oss-20b"
```
Comment (prajwal1210) [inline examples/infer.py]: Should not need full path here.

File: examples/infer.py [hunk 5]
```diff
@@ -45,20 +45,11 @@
     enable_profiling = False
 
     # Tokenizer for encoding the prompt
-    tokenizer = AutoTokenizer.from_pretrained(model_id)
+    tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=True, local_files_only=True)
```
Comment (prajwal1210) [inline examples/infer.py]: Why local files only?

File: examples/infer.py [hunk 6]
```diff
@@ -45,20 +45,11 @@
...
-            {
-                "role": "system",
-                "content": system_prompt,
-            },  # not needed for gemma
-            {"role": "user", "content": user_prompt},
-        ]
-        formatted_prompt = tokenizer.apply_chat_template(
-            conversation, add_generation_prompt=True, tokenize=False
-        )
+        formatted_prompt = f"{system_prompt.strip()}\n\n{user_prompt.strip()}\n"
```
Comment (prajwal1210) [inline examples/infer.py]: If this is gpt-oss specific, then we need to remove this. The example file should not change

File: yalis/attention/backend_impl/sdpa_and_flex.py [hunk 7]
```diff
@@ -134,20 +138,43 @@ def rotary_kv_update_sdpa_gen(
...
+        mask = upper_mask & lower_mask
     else:
-        k_cache[b_indices, :, t_indices, :] = k[:, :, 0, :]
-        v_cache[b_indices, :, t_indices, :] = v[:, :, 0, :]
-    mask = build_mask_from_index(token_counter[:B], t_max=k_cache.size(-2))
+        mask = upper_mask
 
     enable_gqa = q.size(1) != k.size(1)
 
+    # For GQA, expand K and V to match Q's head count
```
Comment (prajwal1210) [inline yalis/attention/backend_impl/sdpa_and_flex.py:169]: This does not seem right. Our implementation had GQA support before. In fact, we have tested SDPA with Llama3-8B and that requires GQA.

**Review by prajwal1210 — CHANGES_REQUESTED** [review]

File: yalis/tensor_parallel/linear.py (hunk 3 above)
Comment (hoffmann-muki) [reply yalis/tensor_parallel/linear.py]: these changes were made to support sharding the checkpoint which I'm no longer going ahead with so I'll revert them

File: examples/infer.py (hunk 2 above)
Comment (hoffmann-muki) [reply examples/infer.py]: alright I'll add a TODO comment there

File: examples/infer.py (hunk 5 above)
Comment (hoffmann-muki) [reply examples/infer.py]: the code tries to fetch the tokenizer from huggingface hub, and since i was working on an offline compute node the program could not run, so local_files_only ensure we use the tokenizer already present on the filesystem

File: yalis/attention/backend_impl/sdpa_and_flex.py (hunk 7 above)
Comment (hoffmann-muki) [reply yalis/attention/backend_impl/sdpa_and_flex.py:169]: The previous implementation was not wrong but it was limited and could results in bugs. GQA support in `torch.nn.functional.scaled_dot_product_attention` is an experimental feature in PyTorch: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html that only works when PyTorch decides to use flash attention or math as attention backend kernels. So if PyTorch decides to use cuDNN or CUTLASS or MPS backends GQA will fail silently or produce wrong results. The current version ensures that regardless of the backend PyTorch decides to use we are correctly computing GQA.

Image

File: tests/basic_correctness/gpt-oss/conftest.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+def alpaca_dataset():
+    """Create an Alpaca dataset for testing."""
+    dataset = AlpacaDataset(random_seed=42)
+    return dataset
+
+
+# Model fixtures - return lazy loaders, not pre-loaded models
+@pytest.fixture(scope="function")
+def hf_model_loader(model_id, dtype, attn_backend, device):
+    """Return a callable that loads HF model on demand (rank 0 only)."""
```
Comment (prajwal1210) [inline tests/basic_correctness/gpt-oss/conftest.py]: Won't this be extremely slow? I am loading the model everytime for each test?

File: tests/basic_correctness/gpt-oss/conftest.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    return request.config.getini("model")
+
+
+@pytest.fixture(scope="module")
+def draft_model_id(request):
+    """Get the draft model ID from pytest config."""
+    return request.config.getini("draft_model")
+
+
+@pytest.fixture(scope="module")
```
Comment (prajwal1210) [inline tests/basic_correctness/gpt-oss/conftest.py]: Why do we need to replicate these in a different folder? Looks like a lot of code is copied from the main conftest.py

File: yalis/attention/backend_impl/sdpa_and_flex.py [hunk 10]
```diff
@@ -174,20 +201,290 @@ def rotary_kv_update_sdpa_gen(
...
+            # Sliding window: query i attends to keys max(0, i-window)..i
+            # Create band matrix
+            row_indices = torch.arange(T, device=q.device).view(T, 1)
+            col_indices = torch.arange(T, device=q.device).view(1, T)
+            window_mask = (row_indices - col_indices) <= sliding_window
+            mask = causal_mask & window_mask
+        else:
+            mask = causal_mask
+        # mask shape: (T, T) - will broadcast to (B, 1, T, T) later
+        n_k = T
```
[the remark this answers is not in the document]
Comment (hoffmann-muki) [reply yalis/attention/backend_impl/sdpa_and_flex.py:436]: it's used in a later context

File: yalis/external/model.py [hunk 11]
```diff
@@ -1057,3 +1175,171 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        mlp2_weight = self.mlp2_weight[expert_indices, ...]  # (M, k, C, per)
+        mlp2_bias = self.mlp2_bias[expert_indices, ...]  # (M, k, C)
+        down = torch.einsum("m k c p, m k p -> m k c", mlp2_weight, up)
+
+        # distributed aggregation when sharded across ranks
+        if torch.distributed.is_available() and torch.distributed.is_initialized():
+            torch.distributed.all_reduce(down, op=torch.distributed.ReduceOp.SUM)
+
+        down = down + mlp2_bias
+
```
[the remark this answers is not in the document]
Comment (hoffmann-muki) [reply yalis/external/model.py]: GPT-OSS reference impl does all reduce before adding the bias

**Comment by hoffmann-muki** [comment]
addressed PR comments raised and linting issues. Now I need to run gpt oss unit tests on 4 GPUs to verify there are no memory issues

File: yalis/model_loading.py [hunk 12]
```diff
@@ -175,6 +175,19 @@ def _load_key(self, name: str) -> torch.Tensor:
...
             t = t.to(device=dev, non_blocking=non_blocking)
+        try:
+            _rank = torch.distributed.get_rank()
+        except Exception:
+            _rank = -1
+        try:
+            shard_info = f"{shard}"
+        except Exception:
+            shard_info = shard
+        print_rank0(
```
Comment (prajwal1210) [inline yalis/model_loading.py]: This seems like debugging info. Remove or make it optional so it doesn't get printed always.

File: yalis/initialize.py [hunk 13]
```diff
@@ -10,11 +10,28 @@ def init_distributed(tp_dims=None):
     # - around 1GB This causes OOMs for some runs. For now, we are not passing
     # device_id which leads to a warning but it works fine. Ideally, we need to
     # find a way to do this to avoid the warning without OOMs.
-    dist.init_process_group(backend="nccl")
+    if not dist.is_initialized():
```
Comment (prajwal1210) [inline yalis/initialize.py]: This seems to be changed for the tests but is not needed for usability. Let me know if otherwise, else remove.

File: yalis/engine.py [hunk 14]
```diff
@@ -236,7 +236,19 @@ def _initialize_model(self, model_config, inference_config):
         print_rank0(
             f"Memory Stats after Symm Pool Creation - {get_gpu_memory_info()} "
         )
-        tokenizer = AutoTokenizer.from_pretrained(model_config.model_name)
+        # Try loading tokenizer from model_path first for local checkpoints,
+        # then fall back to model_name for HF models
+        is_local_model_path = os.path.isdir(model_config.model_path)
```
Comment (prajwal1210) [inline yalis/engine.py]: Is this needed? Why cannot we run GPT-OSS like other models without local paths.

File: yalis/tensor_parallel/linear.py [hunk 15]
```diff
@@ -67,6 +67,34 @@ def divide(a, b):
     return a // b
 
 
+@torch.no_grad()
```
Comment (prajwal1210) [inline yalis/tensor_parallel/linear.py:70]: Is this function being called anywhere?

**Review by prajwal1210 — CHANGES_REQUESTED** [review]

File: yalis/model_loading.py (hunk 12 above)
Comment (hoffmann-muki) [reply yalis/model_loading.py]: alright will take it out

File: yalis/engine.py (hunk 14 above)
Comment (hoffmann-muki) [reply yalis/engine.py]: we can do it, I added this because I was running the code on an offline compute node and by default the code attempts to fetch the tokenizer from the remote repository instead of using that on the filesystem. once I'm done testing I'll revert the change

File: yalis/tensor_parallel/linear.py (hunk 15 above)
Comment (hoffmann-muki) [reply yalis/tensor_parallel/linear.py:70]: yes this is called in the gpt oss moe _modified_load_from_state_dict method for auto sharding when we enable tp

File: tests/basic_correctness/gpt-oss/conftest.py (hunk 8 above)
Comment (hoffmann-muki) [reply tests/basic_correctness/gpt-oss/conftest.py]: yeah these will be removed when i verify the tests run on the regular test suite using 4 GPUs

File: tests/basic_correctness/gpt-oss/conftest.py (hunk 9 above)
Comment (hoffmann-muki) [reply tests/basic_correctness/gpt-oss/conftest.py]: yeah these will be removed when i verify the tests run on the regular test suite using 4 GPUs

File: yalis/tensor_parallel/linear.py (hunk 15 above)
Comment (hoffmann-muki) [reply yalis/tensor_parallel/linear.py:70]: added comments clarifying its usage both here and in model.py where GptOssMoE resides

**Comment by hoffmann-muki** [comment]
Fixed formatting issues, now the PR is ready for review

**Comment by prajwal1210** [comment]
@codex Review

File: yalis/external/model.py [hunk 16]
```diff
@@ -1016,3 +1161,200 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
         y = self.experts(x, router)
         return y.view(B, T, C)
+
+
+class GptOssMoE(nn.Module):
```
Comment (prajwal1210) [inline yalis/external/model.py:1168]: We should not be completely writing this class from scratch. We have a TPMoE class in yalis/tensor_parallel/moe.py that we should be reusing for this. Also, we should use the fused_experts function in it.

**Review by prajwal1210 — CHANGES_REQUESTED** [review]

**Comment by prajwal1210** [comment]
@codex Review
